qutip/tests/test_metrics.py

```diff
@@ -557,5 +557,52 @@ def case(A, significant=4):
             yield case, rand_super_bcsz(dim)
 
 
+@dnorm_test
+def test_dnorm_sparse_vs_dense():
+    """
+    Metrics: checks that the diamond norm is equal when calculated with memoized dense or sparse settings.
+    """
+    # NB: It might be worth dropping test_dnorm_force_solve, and separating
+    #     into cases for each optimization path.
+    def case(A, significant=4):
+        for force_solve in (False, True):
+            assert_approx_equal(
+                dnorm(A, force_solve=force_solve),
+                dnorm(A, force_solve=force_solve, dense_memoized_solve=False)
+            )
+
+    for dim in (2, 3):
+        for _ in range(10):
+            yield case, rand_super_bcsz(dim)
```

Comment:
We use `pytest` as our runner (that's what gets called when you do `qutip.testing.run()`).  So if you want to try out a test run of the file `test_metrics.py` and limit the run to only tests that have `dnorm` in the name, it would look like:
```text
$ pytest qutip/tests/test_metrics.py -k dnorm
======================================== test session starts ========================================
platform darwin -- Python 3.8.8, pytest-6.2.2, py-1.10.0, pluggy-0.13.1
rootdir: /Users/jake/code/qutip/qutip/qutip/tests, configfile: pytest.ini
collected 29 items / 21 deselected / 8 selected

qutip/tests/test_metrics.py xxxxxxxx                                                          [100%]

========================================= warnings summary ==========================================
../../../.anaconda3/envs/qutip-dev/lib/python3.8/site-packages/pyximport/pyximport.py:51
  /Users/jake/.anaconda3/envs/qutip-dev/lib/python3.8/site-packages/pyximport/pyximport.py:51: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
    import imp

test_metrics.py:362
  test_metrics.py:362: PytestCollectionWarning: yield tests were removed in pytest 4.0 - test_unitarity_known will be ignored

test_metrics.py:377
  test_metrics.py:377: PytestCollectionWarning: yield tests were removed in pytest 4.0 - test_unitarity_bounded will be ignored

test_metrics.py:388
  test_metrics.py:388: PytestCollectionWarning: yield tests were removed in pytest 4.0 - test_dnorm_bounded will be ignored

test_metrics.py:401
  test_metrics.py:401: PytestCollectionWarning: yield tests were removed in pytest 4.0 - test_dnorm_qubit_known_cases will be ignored

test_metrics.py:482
  test_metrics.py:482: PytestCollectionWarning: yield tests were removed in pytest 4.0 - test_dnorm_qubit_scalar will be ignored

test_metrics.py:501
  test_metrics.py:501: PytestCollectionWarning: yield tests were removed in pytest 4.0 - test_dnorm_qubit_triangle will be ignored

test_metrics.py:519
  test_metrics.py:519: PytestCollectionWarning: yield tests were removed in pytest 4.0 - test_dnorm_force_solve will be ignored

test_metrics.py:542
  test_metrics.py:542: PytestCollectionWarning: yield tests were removed in pytest 4.0 - test_dnorm_cptp will be ignored

test_metrics.py:560
  test_metrics.py:560: PytestCollectionWarning: yield tests were removed in pytest 4.0 - test_dnorm_sparse_vs_dense will be ignored

test_metrics.py:579
  test_metrics.py:579: PytestCollectionWarning: yield tests were removed in pytest 4.0 - test_dnorm_on_sparse_matrix will be ignored

-- Docs: https://docs.pytest.org/en/stable/warnings.html
========================== 21 deselected, 8 xfailed, 11 warnings in 0.37s ===========================
```

Reply:
Thank you. Should I start a new PR to correct the rest of the metric tests? or is it okey if I do in in this one?

Reply:
Here I'd say the best idea is to write your new tests in the pytest style within this PR and leave the pre-existing ones along for now. You can make another PR that converts the other tests to pytest style if you'd like.

You can look at (for example) https://github.com/qutip/qutip/blob/c1e05a747e0410910c227b593db9c452c13d8024/qutip/tests/test_dimensions.py#L97-L106 to get an idea of the style you write pytest tests in.  The general principle is that you use `pytest.mark.parametrize` instead of `yield`, and you mostly just need to use the Python keyword `assert` rather than special methods.  You can use `pytest.approx` to do floating-point comparisons.

Reply:
I updated all diamond norm tests, I could probably reduce the number of tests even more, but we will have longer and longer lists of parameters.

Reply:
Tests on the ci seem to be breaking because cvxpy installation is optional, should we add it to requirements?

Reply:
We don't want to add `cvxpy` as a full dependency to QuTiP because it's much too heavy and too specialised.  Instead, you should mark your tests as requiring it.  In order to work properly with `pytest`, you'll need to change the lines https://github.com/qutip/qutip/blob/93d587495fbeb6970c324e33744e709f3ffe21a3/qutip/tests/test_metrics.py#L76-L80
to be something more like
```python
if qutip.settings.has_mkl:
    dnorm_test = pytest.mark.xfail(reason='known cvxpy/MKL incompatibility')
else:
    dnorm_test = pytest.mark.skipif(cvxpy is None, reason='requires cvxpy')
```

and then you can decorate your tests with `@dnorm_test`, similar to how the other dnorm tests in that file.

We'll also need to ensure that at least one test runner tests these.  Add `cvxpy` onto the end of this line https://github.com/qutip/qutip/blob/f7bc01bc19b88a756bf3c4863ab14226c48212aa/.travis.yml#L71

---

Don't worry about the parameter lists much - as they are right now, I'm fine to accept them.  However, if you're interested, excessively long parameter lists are often because the parametrisation isn't done quite right.  `pytest` lets you parametrise values separately, so
```python
@pytest.mark.parametrize('a', [1, 2, 3])
@pytest.mark.parametrize('b', [4, 5, 6])
def test_ab(a, b):
    assert True
```
would run the test 9 times in total - the values (a, b) would be (1, 4), (1, 5), (1, 6), (2, 4), (2, 5), and so on.  Another trick is to test by comparison to a reference implementation.  For example, if you have an analytic form for the diamond norm in certain cases, let's call it a function `overrotation_expected(variable)`, then the parametrisation could be
```python
@pytest.mark.parametrize('variable', [1e-3, 3.1e-2, 1e-3, 3.1e-2])
@pytest.mark.parametrize(['reference', 'testee'], [
    pytest.param(overration_expected, overration, id='overrotation'),
    pytest.param(mixture_expected, mixture, id='mixture'),
])
def test_dnorm_known(variable, reference, testee):
    expected = reference(variable)
    id_channel = qutip.to_choi(qutip.qeye(2))
    test = dnorm(testee(variable), id_channel, abs=1e-7)
    assert test == pytest.approx(expected, rel=1e-5)
```

Note also that you can pass functions as parametrised variables - you don't need to "switch" on strings.

Don't worry about this bit, though - I'll accept it as it is because there's lots more changes that we want to make within the tests anyway.

---

Do remember to delete any code (e.g. tests) that you have obsoleted - don't leave dead code lying around.

Reply:
`cvxpy` is not showing any incompatibility with 'mkl', at least from what I have checked localy.  So I changed the definition of 'dnorm_test'  it appears ` Issue #484 is no longer relevant. So in this case it makes sense to test on the environment with `mkl` as you suggested . If we are still unsure about the issue we can test in a `no-mkl` env as well.

Reply:
Oh, I'd forgotten about that old MKL incompatibility.  Thanks for checking it, it's good to try and keep these things up-to-date.  My reason for choosing that particular test was actually just because that's the one we run the coverage analysis on.